File: site/js/bcb.js
/* Fetcher Banco Central — séries temporais do SGS.
 * Par de fontes/bcb.py.
 *
 * IDs verificados em jun/2026. Alterar somente após re-verificar na API:
 *   https://api.bcb.gov.br/dados/serie/bcdata.sgs.{codigo}/dados
 *
 * Difusão IPCA-15:
 *   Não existe série SGS própria (verificado jun/2026).
 *   Calculada a partir dos subitens da tabela IBGE 7062 (parcela com var > 0).
 */
import { arredondar } from "./numeros.js";
import { calcularDifusao } from "./impacto.js";

const BASE_SGS = "https://api.bcb.gov.br/dados/serie/bcdata.sgs.{}/dados";

// Difusão IPCA — série verificada em jun/2026
const SGS_DIFUSAO_IPCA = 21379;

// Núcleos IPCA — variação mensal (%) — IDs verificados em jun/2026
const NUCLEOS = {
  MS: 4466, // Médias aparadas com suavização
  EX0: 11427, // Por exclusão — EX0
  DP: 16122, // Dupla ponderação
  EX3: 27839, // Exclusão EX3
  P55: 28750, // Preços livres — P55
};

// Conjunto para média: MS+EX0+DP+EX3+P55 — atualizado jun/2026
const CONJUNTO_MEDIA = ["MS", "EX0", "DP", "EX3", "P55"];

// ---------------------------------------------------------------------------
// Helpers internos
// ---------------------------------------------------------------------------

/** (dataInicial, dataFinal) dd/mm/aaaa para os 12m encerrados em mesRef. */
function intervalo12m(mesRef) {
  const ano = parseInt(mesRef.slice(0, 4), 10);
  const m = parseInt(mesRef.slice(4), 10);
  let mIni = m - 11;
  let anoIni = ano;
  if (mIni <= 0) {
    mIni += 12;
    anoIni -= 1;
  }
  const dois = (n) => String(n).padStart(2, "0");
  return [`01/${dois(mIni)}/${anoIni}`, `28/${dois(m)}/${ano}`];
}

/** Registros da série SGS no intervalo. Devolve [] em caso de erro. */
async function sgsFetch(codigo, ini, fim) {
  const url =
    BASE_SGS.replace("{}", String(codigo)) +
    `?formato=json&dataInicial=${encodeURIComponent(ini)}` +
    `&dataFinal=${encodeURIComponent(fim)}`;
  try {
    // Timeout espelha o httpx.get(timeout=30) do Python. Sem ele, uma serie
    // pendurada travaria a nota inteira; com ele, a falha vira [] e o bloco
    // correspondente some da nota — que e a politica do projeto.
    const resp = await fetch(url, { signal: AbortSignal.timeout(30000) });
    if (!resp.ok) return [];
    return (await resp.json()) ?? [];
  } catch {
    return [];
  }
}

function parseFloatBr(s) {
  if (s === null || s === undefined) return null;
  const v = Number(String(s).replace(",", "."));
  return Number.isNaN(v) ? null : v;
}

/** Valor da série para mesRef (AAAAMM), ou null se indisponível. */
async function buscarValorMes(codigo, mesRef) {
  const m = mesRef.slice(4);
  const a = mesRef.slice(0, 4);
  const data = `01/${m}/${a}`;
  const dados = await sgsFetch(codigo, data, data);
  if (dados.length === 0) return null;
  return parseFloatBr(dados[0].valor);
}

/**
 * Acumulado em 12 meses (composto) da série SGS encerrado em mesRef.
 * Devolve null se a série tiver menos de 12 observações no período.
 */
async function acum12mSgs(codigo, mesRef) {
  const [ini, fim] = intervalo12m(mesRef);
  const dados = await sgsFetch(codigo, ini, fim);
  if (dados.length < 12) return null;
  let prod = 1.0;
  for (const d of dados) {
    const val = parseFloatBr(d.valor);
    if (val === null) return null;
    prod *= 1 + val / 100;
  }
  return arredondar((prod - 1) * 100, 4);
}

// ---------------------------------------------------------------------------
// API pública
// ---------------------------------------------------------------------------

/**
 * Média aritmética dos acumulados em 12m dos núcleos MS+EX0+DP+EX3+P55.
 *
 * Devolve null se QUALQUER uma das séries falhar. Uma média parcial é pior
 * que a ausência do dado: com 4 de 5 séries o número publicado desloca em até
 * 0,09 p.p. (ex.: 12m até mar/2026 = 4,39% com as 5 séries, 4,44% sem a DP)
 * sem qualquer sinal de erro. O bloco de núcleo é opcional na nota — omiti-lo
 * é seguro, publicá-lo errado não é.
 */
export async function mediaNucleos12m(mesRef) {
  // NÃO paralelizar com Promise.all. Já foi tentado: disparar as 5 séries de
  // uma vez (12 requisições simultâneas somando os dois meses) faz o SGS
  // derrubar parte delas, e a nota sai sem a comparação com o mês anterior.
  // O ganho seria ~1s; o custo é perder um número. Mantém-se em série.
  const vals = [];
  for (const nome of CONJUNTO_MEDIA) {
    const v = await acum12mSgs(NUCLEOS[nome], mesRef);
    if (v === null) return null;
    vals.push(v);
  }
  const soma = vals.reduce((a, b) => a + b, 0);
  return arredondar(soma / vals.length, 4);
}

/**
 * Preenche difusao, difusao_anterior, nucleo_12m e nucleo_12m_anterior no
 * ResultadoInflacao fornecido. Falhas deixam os campos em null.
 *
 * IPCA:
 *   difusao / difusao_anterior → SGS 21379
 *   nucleo_12m / nucleo_12m_anterior → média MS+EX0+DP+EX3+P55 em 12m
 *
 * IPCA-15:
 *   difusao → calculada dos subitens (nivel 4) já presentes em resultado
 *   difusao_anterior → null (requer fetch extra do IBGE; não implementado)
 *   nucleo_12m → null (BCB não publica núcleo do IPCA-15)
 */
export async function enriquecer(resultado) {
  if (resultado.indicador === "IPCA") {
    // Em série, pelo mesmo motivo descrito em mediaNucleos12m: o SGS não
    // tolera a rajada e passa a devolver menos dados do que existe.
    resultado.difusao = await buscarValorMes(SGS_DIFUSAO_IPCA, resultado.mes_ref);
    resultado.difusao_anterior = await buscarValorMes(
      SGS_DIFUSAO_IPCA,
      resultado.mes_ant,
    );
    resultado.nucleo_12m = await mediaNucleos12m(resultado.mes_ref);
    resultado.nucleo_12m_anterior = await mediaNucleos12m(resultado.mes_ant);
  } else {
    resultado.difusao = calcularDifusao(resultado.subitens, 4);
    resultado.difusao_anterior = null;
    resultado.nucleo_12m = null;
    resultado.nucleo_12m_anterior = null;
  }
}

File: fontes/bcb.py
"""
Fetcher Banco Central — SGS séries temporais.

IDs verificados em jun/2026. Alterar somente após re-verificar na API:
  https://api.bcb.gov.br/dados/serie/bcdata.sgs.{codigo}/dados

Combinação para média dos núcleos (MS+EX0+DP+EX3+P55):
  Atualizado jun/2026: 5 séries — 4466, 11427, 16122, 27839, 28750.
  Requer nova validação contra referência conhecida antes de marcar como estável.

Difusão IPCA-15:
  Não existe série SGS própria (verificado jun/2026).
  Calculada a partir dos subitens da tabela IBGE 7062 (parcela com var > 0).
  Validada: 65,12% (mai/2026) vs. 65,1% publicado — dif 0,02 p.p.
"""
import httpx
from typing import Optional

from .modelo import ItemInflacao, ResultadoInflacao

_BASE_SGS = "https://api.bcb.gov.br/dados/serie/bcdata.sgs.{}/dados"

# Difusão IPCA — série verificada em jun/2026
_SGS_DIFUSAO_IPCA = 21379

# Núcleos IPCA — variação mensal (%) — IDs verificados em jun/2026
_NUCLEOS: dict[str, int] = {
    "MS":  4466,    # Médias aparadas com suavização
    "EX0": 11427,   # Por exclusão — EX0
    "DP":  16122,   # Dupla ponderação
    "EX3": 27839,   # Exclusão EX3
    "P55": 28750,   # Preços livres — P55
}

# Conjunto para média: MS+EX0+DP+EX3+P55 — atualizado jun/2026
_CONJUNTO_MEDIA: tuple[str, ...] = ("MS", "EX0", "DP", "EX3", "P55")


# ---------------------------------------------------------------------------
# Helpers internos
# ---------------------------------------------------------------------------

def _intervalo_12m(mes_ref: str) -> tuple[str, str]:
    """Retorna (dataInicial, dataFinal) dd/mm/aaaa para os 12m encerrados em mes_ref."""
    ano, m = int(mes_ref[:4]), int(mes_ref[4:])
    m_ini = m - 11
    ano_ini = ano
    if m_ini <= 0:
        m_ini += 12
        ano_ini -= 1
    return f"01/{m_ini:02d}/{ano_ini}", f"28/{m:02d}/{ano}"


def _sgs_fetch(codigo: int, ini: str, fim: str) -> list[dict]:
    """
    Busca registros da série SGS no intervalo de datas.
    Retorna [] silenciosamente em caso de erro ou série sem dados.
    """
    url = _BASE_SGS.format(codigo)
    try:
        resp = httpx.get(
            url,
            params={"formato": "json", "dataInicial": ini, "dataFinal": fim},
            timeout=30,
        )
        if resp.status_code != 200:
            return []
        return resp.json() or []
    except Exception:
        return []


def _parse_float(s) -> Optional[float]:
    try:
        return float(str(s).replace(",", "."))
    except (ValueError, TypeError):
        return None


def _buscar_valor_mes(codigo: int, mes_ref: str) -> Optional[float]:
    """Retorna o valor da série para mes_ref (AAAAMM), ou None se indisponível."""
    m, a = mes_ref[4:], mes_ref[:4]
    ini = fim = f"01/{m}/{a}"
    dados = _sgs_fetch(codigo, ini, fim)
    if not dados:
        return None
    return _parse_float(dados[0].get("valor"))


def _acum12m_sgs(codigo: int, mes_ref: str) -> Optional[float]:
    """
    Compõe o acumulado em 12 meses (composto) da série SGS encerrado em mes_ref.
    Retorna None se a série tiver menos de 12 observações no período.
    """
    ini, fim = _intervalo_12m(mes_ref)
    dados = _sgs_fetch(codigo, ini, fim)
    if len(dados) < 12:
        return None
    prod = 1.0
    for d in dados:
        val = _parse_float(d.get("valor"))
        if val is None:
            return None
        prod *= 1 + val / 100
    return round((prod - 1) * 100, 4)


def _difusao_from_subitens(
    subitens: list[ItemInflacao],
    nivel_min: int = 4,
) -> Optional[float]:
    """
    Parcela de subitens (nivel >= nivel_min) com variação positiva no mês.
    Usado para difusão do IPCA-15 na ausência de série SGS própria.
    """
    candidatos = [s for s in subitens if s.nivel >= nivel_min]
    if not candidatos:
        return None
    positivos = sum(1 for s in candidatos if s.variacao > 0)
    return round(positivos / len(candidatos) * 100, 2)


# ---------------------------------------------------------------------------
# API pública
# ---------------------------------------------------------------------------

def media_nucleos_12m(mes_ref: str) -> Optional[float]:
    """
    Média aritmética dos acumulados em 12m dos núcleos MS+EX0+DP+EX3+P55.

    Retorna None se QUALQUER uma das séries falhar. Uma média parcial é pior
    que a ausência do dado: com 4 de 5 séries o número publicado desloca em
    até 0,09 p.p. (ex.: 12m até mar/2026 = 4,39% com as 5 séries, 4,44% sem a
    DP) sem qualquer sinal de erro. O bloco de núcleo é opcional na nota —
    omiti-lo é seguro, publicá-lo errado não é.
    """
    vals = []
    for nome in _CONJUNTO_MEDIA:
        v = _acum12m_sgs(_NUCLEOS[nome], mes_ref)
        if v is None:
            return None
        vals.append(v)
    return round(sum(vals) / len(vals), 4)


def enriquecer(resultado: ResultadoInflacao) -> None:
    """
    Preenche difusao, difusao_anterior, nucleo_12m e nucleo_12m_anterior
    no ResultadoInflacao fornecido. Falhas são silenciosas (campos ficam None).

    IPCA:
      difusao / difusao_anterior → SGS 21379
      nucleo_12m / nucleo_12m_anterior → média MS+EX0+DP+EX3+P55 em 12m

    IPCA-15:
      difusao → calculada dos subitens (nivel 4) já presentes em resultado
      difusao_anterior → None (requer fetch extra do IBGE; não implementado)
      nucleo_12m → None (BCB não publica núcleo do IPCA-15)
    """
    if resultado.indicador == "IPCA":
        resultado.difusao = _buscar_valor_mes(_SGS_DIFUSAO_IPCA, resultado.mes_ref)
        resultado.difusao_anterior = _buscar_valor_mes(_SGS_DIFUSAO_IPCA, resultado.mes_ant)
        resultado.nucleo_12m = media_nucleos_12m(resultado.mes_ref)
        resultado.nucleo_12m_anterior = media_nucleos_12m(resultado.mes_ant)
    else:
        resultado.difusao = _difusao_from_subitens(resultado.subitens, nivel_min=4)
        resultado.difusao_anterior = None
        resultado.nucleo_12m = None
        resultado.nucleo_12m_anterior = None
